File: py/portfolio.py
import functools
import operator

from money import Money

class Portfolio:
    def __init__(self):
        self.moneys = []

    def __convert(self, aMoney, aCurrency):
        exchangeRates = {'EUR->USD': 1.2, 'USD->KRW': 1100}

        if aMoney.currency == aCurrency:
            return aMoney.amount
        
        key = aMoney.currency + '->' + aCurrency
        return aMoney.amount * exchangeRates[key]

    def add(self, *moneys):
        self.moneys.extend(moneys)

    def evaluate(self, bank, currency):
        total = 0.0
        failures = []
        for m in self.moneys:
            try:
                total += bank.convert(m, currency).amount
            except Exception as ex:
                failures.append(ex)

        if not failures:
            return Money(total, currency)

        failureMessage = failureMessage = ",".join(f.args[0] for f in failures)
        raise Exception("Missing exchange rate(s):[" + failureMessage + "]")

File: js/portfolio.js
const Money = require('./money');

class Portfolio {
    constructor() {
        this.moneys = [];
    }

    convert(money, currency) {
        let eurToUsd = 1.2;
        if(money.currency === currency) {
            return money.amount;
        }
        return money.amount * eurToUsd;
    }

    add(...moneys) {
        this.moneys = this.moneys.concat(moneys);
    }

    evaluate(currency) {
        let total = this.moneys.reduce((sum, money) => {
            return sum + this.convert(money, currency);
        }, 0);
        return new Money(total, currency);
    }
}

module.exports = Portfolio